### ISSUE TITLE
Ordinary least squares for the Runge function

### ISSUE THREAD

**Issue by SigurdRiseth** [SigurdRiseth/FYS-STK4155-P1#1]
We will generate our own data set for the above function $f(x)$ with $x\in[-1,1]$, using either a uniform distribution or a fixed step size to set up the array of $x$ values.  You should also explore the addition of normally distributed noise $\varepsilon\sim\mathcal{N}(0,\sigma^2)$ to the function; a noise level of $\sigma\approx0.1$ is a good starting point, but study how your conclusions depend on it.

**Write your own code** (using for example the pseudoinverse function `pinv` from `numpy`, or the SVD directly) and perform a standard **ordinary least squares regression** analysis using polynomials in $x$ up to order $15$ or higher.  Explore the dependence on the number of data points and the polynomial degree.

Evaluate the mean squared error (MSE)

$$
\mathrm{MSE}(\mathbf{y},\tilde{\mathbf{y}}) = \frac{1}{n}\sum_{i=0}^{n-1}(y_i-\tilde{y}_i)^2,
$$

and the $R^2$ score function.  If $\tilde{y}_i$ is the predicted value of the $i$-th sample and $y_i$ is the corresponding true value, then the score $R^2$ is defined as

$$
R^2(\mathbf{y},\tilde{\mathbf{y}}) = 1 - \frac{\sum_{i=0}^{n-1}(y_i-\tilde{y}_i)^2}
                                      {\sum_{i=0}^{n-1}(y_i-\bar{y})^2},
  \qquad
  \bar{y} = \frac{1}{n}\sum_{i=0}^{n-1}y_i .
$$

Plot the resulting scores (MSE and $R^2$) as functions of the polynomial degree (here up to polynomial degree 15).  Plot also the parameters $\boldsymbol{\theta}$ as you increase the order of the polynomial.  Comment your results.

Your code has to include a scaling/centering of the data (for example by subtracting the mean value), and a split of the data in training and test
data.  For the splitting you can either write your own code or use for example the function `train_test_split` provided by **Scikit-Learn** (make sure you have installed it).  **You should present a critical discussion of why and how you have scaled/ or not scaled the data** (see Section 3.13 of the lecture notes and the Tuesday session of week 35 and the pertinent notebook for week 35).

It is normal in essentially all Machine Learning studies to split the data in a training set and a test set (eventually also an additional validation set).  There is no explicit recipe for how much data should be included as training data and say test data.  An accepted rule of thumb is to use approximately $2/3$ to $4/5$ of the data as training data.

You can easily reuse the solutions to your exercises from week 35 (the analytical derivatives of the cost function, your own OLS code, and the train/test analysis).  See also the lecture slides from weeks 34 and 35 and Chapter 3 of the lecture notes.  On scaling, we recommend reading the following section from the Scikit-Learn documentation: https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html

See also section 3.13 of the lecture notes.
